### ISSUE TITLE
Fatal Error: failed to open device ens1f1: Failure accessing netmap.

### ISSUE THREAD

**Issue by KatherineJXY** [appneta/tcpreplay#834]
r6@r6-PowerEdge-R740:~$ sudo tcpreplay --netmap -i ens1f1 ./home/r6/scripts-bk/trace/univ1_trace/univ1_pt1
Switching network driver for ens1f1 to netmap bypass mode...
Fatal Error: failed to open device ens1f1: Failure accessing netmap.
        Request for netmap version 14 failed.
        Compiled netmap driver is version 14.
        Error=Invalid argument

I have tried multiple versions of both tcpreplay and netmap, however, I got the same answer. 

More info:
r6@r6-PowerEdge-R740:~$ uname -a
Linux r6-PowerEdge-R740 5.15.0-78-generic #85~20.04.1-Ubuntu SMP Mon Jul 17 09:42:39 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

r6@r6-PowerEdge-R740:~$ tcpreplay -V
Warning: May need to run as root to get access to all network interfaces.
tcpreplay version: 4.4.4 (build git:v4.4.4)
Copyright 2013-2022 by Fred Klassen <tcpreplay at appneta dot com> - AppNeta
Copyright 2000-2012 by Aaron Turner <aturner at synfin dot net>
The entire Tcpreplay Suite is licensed under the GPLv3
Cache file supported: 04
Not compiled with libdnet.
Compiled against libpcap: 1.9.1
64 bit packet counters: enabled
Verbose printing via tcpdump: enabled
Packet editing: disabled
Fragroute engine: disabled
Default injection method: PF_PACKET send()
Optional injection method: netmap

**Comment by KatherineJXY** [comment]
r6@r6-PowerEdge-R740:~/Downloads/netmap/apps$ sudo pkt-gen -i ens1f1 -f tx -b1 -l 1400
618.605778 main [3036] interface is ens1f1
618.605877 main [3167] running on 1 cpus (have 20)
618.606304 extract_ip_range [477] range is 10.0.0.1:1234 to 10.0.0.1:1234
618.606317 extract_ip_range [477] range is 10.1.0.1:1234 to 10.1.0.1:1234
ens1f1: Invalid argument
618.607837 main [3335] aborting

I got the same error when I used pkt-gen to generate packets.

**Comment by ChuckCottrill** [comment]
I have recently compiled and ran tcpreplay on both Ubuntu and Rocky Linux.
Have you tried with other versions of tcpreplay? Have you tried building on a different OS?
This may be a problem with your network driver and not specifically tcpreplay.
Have you tried installing and running either wireshark or tcpdump?
I do not have a PowerEdge system to try to replicate your issue.

**Comment by KatherineJXY** [comment]
> I have recently compiled and ran tcpreplay on both Ubuntu and Rocky Linux. Have you tried with other versions of tcpreplay? Have you tried building on a different OS? This may be a problem with your network driver and not specifically tcpreplay. Have you tried installing and running either wireshark or tcpdump? I do not have a PowerEdge system to try to replicate your issue.

Yes, it seems like this problem comes from netmap. I can use tcpreplay to send packets without netmap. I changed the network driver from mlx5 to ice, and it can work well now.

**Comment by fklassen** [comment]
Closing - this is external to tcpreplay, not a bug here.

You already proved it yourself: running netmap's own `pkt-gen` tool against the same `ens1f1` interface reproduces the identical `Invalid argument` error, with zero tcpreplay code involved. That confirms the failure is in the netmap kernel module / mlx5 driver's interface-registration path (the `NIOCREGIF` ioctl), not in tcpreplay's netmap integration.

Checked the code on our side: `sendpacket_open_netmap()` (src/common/netmap.c) reports exactly what the kernel returns for that ioctl - netmap API version, compiled version, and `strerror(errno)`. That's already correct and as much diagnostic detail as userspace can extract from a kernel-side rejection; there's no logic bug or missing check to fix here.

Your own workaround, for anyone else who lands on this issue with the same combo: switching from the `mlx5` driver/NIC to `ice` resolved it - mlx5's netmap support is less complete than Intel's ixgbe/i40e/ice. Worth trying a different NIC/driver, or filing directly against the netmap project if you need mlx5 support specifically.
